### PR TITLE
Experimental: gas cost estination for `usize`

### PR DESCRIPTION
**Pull request by mrLSD**
## Description

That's the experimental usage of SpuntikVM with direct checking for `usize` casting overflow, which avoids `aurora-engine` panics.

## SputnikVM

Based on experimental PR: aurora-is-near/sputnikvm/pull/49

## Gas costs
- for some tests gas cost increase nearly `1%`
- for some tests gas cost increase nearly `10%`

### CODE REVIEW

File: engine/src/engine.rs [hunk 1]
```diff
@@ -1915,6 +1920,23 @@ impl<'env, I: IO + Copy, E: Env, M: ModExpAlgorithm> Backend for Engine<'env, I,
...
 
+    fn is_empty_storage(&self, address: H160) -> bool {
+        let address = Address::new(address);
+        if self
+            .contract_storage_cache
+            .borrow()
+            .contains_key(&(address, H256::zero()))
+        {
+            return true;
+        }
```
Comment (birchmd) [inline engine/src/engine.rs:1931]: I don't think this implementation is correct. If the storage cache contains a key then it is not empty (should return `false`). Also what if `0x00` is not used, but some other slot is?

File: engine-tests/src/tests/one_inch.rs [hunk 2]
```diff
@@ -26,7 +26,7 @@ fn test_1inch_liquidity_protocol() {
 
     let (result, profile, pool_factory) = helper.create_pool_factory(&deployer_address);
     assert!(result.gas_used >= 2_800_000); // more than 2.8M EVM gas used
-    assert_gas_bound(profile.all_gas(), 9); // less than 9 NEAR Tgas used
+    assert_gas_bound(profile.all_gas(), 10); // less than 9 NEAR Tgas used
```
Comment (birchmd) [inline engine-tests/src/tests/one_inch.rs:29]: Gas costs going up a little is not ideal. But I think it is worth the trade off of not having panics in sputnikvm.

File: engine/src/engine.rs (hunk 1 above)
Comment (mrLSD) [reply engine/src/engine.rs:1931]: In that particular case, we're just checking a storage cache.
But if it doesn't exist we'll just read storage by the key: (address, index), where index = 0;
Index zero should exist, as it's the impossible case in EVM to have index > 0, but don't have an index = 0;

Also, we can't store storage keys for an account, as it'll cost NEAR storage.

File: engine-tests/src/tests/one_inch.rs (hunk 2 above)
Comment (mrLSD) [reply engine-tests/src/tests/one_inch.rs:29]: In that case, it's `10%`. If you think it's normal - then no problem.

But also, we should charge gas for non-validation errors, from execution.

File: engine/src/engine.rs (hunk 1 above)
Comment (birchmd) [reply engine/src/engine.rs:1931]: > it's the impossible case in EVM to have index > 0, but don't have an index = 0

Why is this the case? I thought you could write a value to any slot leaving the others unused. The storage usage convention in Solidity contracts is to store values for fields in the contract starting at index 0, but for arbitrary contracts written in EVM bytecode directly that does not need to be the case.

Comment (mrLSD) [reply engine/src/engine.rs:1931]: hm... it's problematic to get account storage status on `aurora-engine`. We don't know the account storage keys (indexes) for current EVM accounts. I believe it's the optimal solution for now to assume that `inxed=0` exists.
In the future, we can store account empty storage flag.

Comment (birchmd) [reply engine/src/engine.rs:1931]: Yeah, we'll need to think about a more robust solution to this problem. A flag is a nice idea. It'll require a storage migration for all existing accounts, but it's probably worth the effort to do this.

**Pull request closed without merging**